database/rrdhost.c

```diff
@@ -811,82 +811,66 @@ void rrdhost_cleanup_obsolete_charts(RRDHOST *host) {
 // ----------------------------------------------------------------------------
 // RRDHOST - set system info from environment variables
 
-int rrdhost_set_system_info_variable(struct rrdhost_system_info *system_info, char *name, char *value) {
+int rrdhost_set_system_info_variable(RRDHOST *host, struct rrdhost_system_info *system_info, char *name, char *value) {
+    if(host) rrdhost_wrlock(host);
+
     if(!strcmp(name, "NETDATA_SYSTEM_OS_NAME")){
+        freez(system_info->os_name);
         system_info->os_name = strdupz(value);
     }
     else if(!strcmp(name, "NETDATA_SYSTEM_OS_ID")){
+        freez(system_info->os_id);
         system_info->os_id = strdupz(value);
     }
     else if(!strcmp(name, "NETDATA_SYSTEM_OS_ID_LIKE")){
+        freez(system_info->os_id_like);
         system_info->os_id_like = strdupz(value);
     }
     else if(!strcmp(name, "NETDATA_SYSTEM_OS_VERSION")){
+        freez(system_info->os_version);
         system_info->os_version = strdupz(value);
     }
     else if(!strcmp(name, "NETDATA_SYSTEM_OS_VERSION_ID")){
+        freez(system_info->os_version_id);
         system_info->os_version_id = strdupz(value);
     }
     else if(!strcmp(name, "NETDATA_SYSTEM_OS_DETECTION")){
+        freez(system_info->os_detection);
         system_info->os_detection = strdupz(value);
     }
     else if(!strcmp(name, "NETDATA_SYSTEM_KERNEL_NAME")){
+        freez(system_info->kernel_name);
         system_info->kernel_name = strdupz(value);
     }
     else if(!strcmp(name, "NETDATA_SYSTEM_KERNEL_VERSION")){
+        freez(system_info->kernel_version);
         system_info->kernel_version = strdupz(value);
     }
     else if(!strcmp(name, "NETDATA_SYSTEM_ARCHITECTURE")){
+        freez(system_info->architecture);
         system_info->architecture = strdupz(value);
     }
     else if(!strcmp(name, "NETDATA_SYSTEM_VIRTUALIZATION")){
+        freez(system_info->virtualization);
         system_info->virtualization = strdupz(value);
     }
     else if(!strcmp(name, "NETDATA_SYSTEM_VIRT_DETECTION")){
+        freez(system_info->virt_detection);
         system_info->virt_detection = strdupz(value);
     }
     else if(!strcmp(name, "NETDATA_SYSTEM_CONTAINER")){
+        freez(system_info->container);
         system_info->container = strdupz(value);
     }
     else if(!strcmp(name, "NETDATA_SYSTEM_CONTAINER_DETECTION")){
+        freez(system_info->container_detection);
         system_info->container_detection = strdupz(value);
     }
-    else return 1;
+    else {
+        if(host) rrdhost_unlock(host);
+        return 1;
```

Comment:
can i propose here to use a variable  "int RESULT=0" that you will set to 1 at this point.
Then you only do in the end the unlock and "return RESULT";